### ISSUE TITLE
Segmentation fault in ossec-logcollector

### ISSUE THREAD

**Issue by ssigwart** [wazuh/wazuh#3992]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 3.10.0 and 3.10.1 | ossec-logcollector | Agent | Wazuh Repo | Amazon Linux 2 |

`ossec-logcollector` runs for a few minutes, then crashes.  I tried setting `ossec-logcollector.debug=2`, but it didn't output anything useful when it crashed.

The issue seems to have started with 3.10.0 and 3.10.1 had the same issue.  I downgraded to 3.9.5 and don't have the issue.

### /var/log/messages
```
kernel: ossec-logcollec[31186]: segfault at 0 ip           (null) sp 00007fbc383ddda8 error 14 in ossec-logcollector[400000+b8000]
```

### ossec.conf
```
<ossec_config>
  <client>
    <server>
      <address>XXXXXX</address>
      <port>1514</port>
      <protocol>udp</protocol>
    </server>
    <config-profile>amzn, amzn2018, amzn2018.03</config-profile>
    <notify_time>10</notify_time>
    <time-reconnect>60</time-reconnect>
    <auto_restart>yes</auto_restart>
    <crypto_method>aes</crypto_method>
  </client>

  <client_buffer>
    <!-- Agent buffer options -->
    <disabled>no</disabled>
    <queue_size>10000</queue_size>
    <events_per_second>1000</events_per_second>
  </client_buffer>

  <!-- Policy monitoring -->
  <rootcheck>
    <disabled>no</disabled>
    <check_unixaudit>yes</check_unixaudit>
    <check_files>yes</check_files>
    <check_trojans>yes</check_trojans>
    <check_dev>yes</check_dev>
    <check_sys>yes</check_sys>
    <check_pids>yes</check_pids>
    <check_ports>yes</check_ports>
    <check_if>yes</check_if>

    <!-- Frequency that rootcheck is executed - every 12 hours -->
    <frequency>43200</frequency>

    <rootkit_files>/var/ossec/etc/shared/rootkit_files.txt</rootkit_files>
    <rootkit_trojans>/var/ossec/etc/shared/rootkit_trojans.txt</rootkit_trojans>

    <system_audit>/var/ossec/etc/shared/system_audit_rcl.txt</system_audit>
    <system_audit>/var/ossec/etc/shared/system_audit_ssh.txt</system_audit>

    <skip_nfs>yes</skip_nfs>
  </rootcheck>

  <wodle name="open-scap">
    <disabled>yes</disabled>
    <timeout>1800</timeout>
    <interval>1d</interval>
    <scan-on-start>yes</scan-on-start>
  </wodle>

  <wodle name="cis-cat">
    <disabled>yes</disabled>
    <timeout>1800</timeout>
    <interval>1d</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>wodles/java</java_path>
    <ciscat_path>wodles/ciscat</ciscat_path>
  </wodle>

  <!-- Osquery integration -->
  <wodle name="osquery">
    <disabled>yes</disabled>
    <run_daemon>yes</run_daemon>
    <log_path>/var/log/osquery/osqueryd.results.log</log_path>
    <config_path>/etc/osquery/osquery.conf</config_path>
    <add_labels>yes</add_labels>
  </wodle>

  <!-- System inventory -->
  <wodle name="syscollector">
    <disabled>no</disabled>
    <interval>1h</interval>
    <scan_on_start>yes</scan_on_start>
    <hardware>yes</hardware>
    <os>yes</os>
    <network>yes</network>
    <packages>yes</packages>
    <ports all="no">yes</ports>
    <processes>yes</processes>
  </wodle>

  <!-- File integrity monitoring -->
  <syscheck>
    <disabled>no</disabled>
    <!-- Frequency that syscheck is executed -- default every 2 hours -->
    <frequency>7200</frequency>

    <scan_on_start>yes</scan_on_start>

    <!-- Directories to check  (perform all possible verifications) -->
    <directories check_all="yes">/etc,/usr/bin,/usr/sbin</directories>
    <directories check_all="yes">/bin,/sbin,/boot</directories>

    <!-- Files/directories to ignore -->
    <ignore>/etc/motd</ignore>
    <ignore>/etc/mtab</ignore>
    <ignore>/etc/hosts.deny</ignore>
    <ignore>/etc/mail/statistics</ignore>
    <ignore>/etc/random-seed</ignore>
    <ignore>/etc/random.seed</ignore>
    <ignore>/etc/adjtime</ignore>
    <ignore>/etc/httpd/logs</ignore>
    <ignore>/etc/nginx/cache</ignore>
    <ignore>/etc/utmpx</ignore>
    <ignore>/etc/wtmpx</ignore>
    <ignore>/etc/cups/certs</ignore>
    <ignore>/etc/dumpdates</ignore>
    <ignore>/etc/svc/volatile</ignore>
    <ignore>/sys/kernel/security</ignore>
    <ignore>/sys/kernel/debug</ignore>

    <!-- Check the file, but never compute the diff -->
    <nodiff>/etc/ssl/private.key</nodiff>

    <skip_nfs>yes</skip_nfs>

    <!-- Remove not monitored files -->
    <remove_old_diff>yes</remove_old_diff>
  </syscheck>

  <rootcheck>
    <rootkit_files>/var/ossec/etc/shared/rootkit_files.txt</rootkit_files>
    <rootkit_trojans>/var/ossec/etc/shared/rootkit_trojans.txt</rootkit_trojans>
  </rootcheck>

  <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/messages</location>
  </localfile>

  <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/authlog</location>
  </localfile>

  <localfile>
    <log_format>audit</log_format>
    <location>/var/log/audit/audit.log</location>
  </localfile>

  <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/secure</location>
  </localfile>

  <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/xferlog</location>
  </localfile>

  <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/maillog</location>
  </localfile>

  <localfile>
    <log_format>command</log_format>
    <command>df -P</command>
    <frequency>360</frequency>
  </localfile>

  <localfile>
    <log_format>full_command</log_format>
    <command>netstat -tulpn | sed 's/\([[:alnum:]]\+\)\ \+[[:digit:]]\+\ \+[[:digit:]]\+\ \+\(.*\):\([[:digit:]]*\)\ \+\([0-9\.\:\*]\+\).\+\ \([[:digit:]]*\/[[:alnum:]\-]*\).*/\1 \2 == \3 == \4 \5/' | sort -k 4 -g | sed 's/ == \(.*\) ==/:\1/' | sed 1,2d</command>
    <alias>netstat listening ports</alias>
    <frequency>360</frequency>
  </localfile>

  <localfile>
    <log_format>full_command</log_format>
    <command>last -n 20</command>
    <frequency>360</frequency>
  </localfile>

  <localfile>
    <log_format>command</log_format>
    <frequency>604800</frequency>
    <command>/usr/bin/yum --disableplugin=versionlock --debuglevel 2 --security check-update 2&gt;/dev/null | grep -P '(?&lt;! 0 packages) available$'</command>
  </localfile>


  <localfile>
    <log_format>apache</log_format>
    <location>/var/www/logs/access_log</location>
  </localfile>
  <localfile>
    <log_format>apache</log_format>
    <location>/var/www/logs/error_log</location>
  </localfile>
  <localfile>
    <log_format>apache</log_format>
    <location>/var/log/httpd/access_log.%Y%m%d-000000</location>
  </localfile>
  <localfile>
    <log_format>apache</log_format>
    <location>/var/log/httpd/request_log.%Y%m%d-000000</location>
  </localfile>
  <localfile>
    <log_format>apache</log_format>
    <location>/var/log/httpd/error_log.%Y%m%d-000000</location>
  </localfile>
  <localfile>
    <log_format>apache</log_format>
    <location>/var/log/httpd/ssl_access_log.%Y%m%d-000000</location>
  </localfile>
  <localfile>
    <log_format>apache</log_format>
    <location>/var/log/httpd/ssl_request_log.%Y%m%d-000000</location>
  </localfile>
  <localfile>
    <log_format>apache</log_format>
    <location>/var/log/httpd/ssl_error_log.%Y%m%d-000000</location>
  </localfile>




</ossec_config>
```

**Comment by chemamartinez** [comment]
Hi @ssigwart,

Thank you for reporting us this issue. We are trying to reproduce it, could you share us the debug logcollector messages you have in your `ossec.log` file until the crash? It would be useful for us to detect the bug.

Best regards,
Chema.

**Comment by vikman90** [comment]
|Branch|
|---|
|[3.10-fix-logcollector-3992](https://github.com/wazuh/wazuh/compare/3.10-fix-logcollector-3992)|

Hi @ssigwart @chemamartinez.

Just working on this issue. We have detected the cause of this crash: Logcollector was not filling some internal data fields if any of the files configured was missing.

On the other hand, we are reviewing the behavior related to file management in case of having file links.

We will keep you posted.

Thank you again for letting us know, and apologies for the inconveniences.

Best regards.

**Comment by ssigwart** [comment]
Thanks for the quick responses.  Do you have any estimate on when the fix will be released?

Here are some of the logs:
```
2019/09/19 20:26:29 ossec-logcollector: ERROR: Couldn't stat file '/var/log/authlog'
2019/09/19 20:26:29 ossec-logcollector: ERROR: Couldn't stat file '/var/log/authlog'
2019/09/19 20:26:29 ossec-logcollector: ERROR: Couldn't stat file '/var/log/authlog'
2019/09/19 20:26:29 ossec-logcollector: ERROR: Couldn't stat file '/var/log/authlog'
2019/09/19 20:26:29 ossec-logcollector: ERROR: Couldn't stat file '/var/log/xferlog'
2019/09/19 20:26:29 ossec-logcollector: ERROR: Couldn't stat file '/var/log/authlog'
2019/09/19 20:26:29 ossec-logcollector: INFO: Monitoring output of command(360): df -P
2019/09/19 20:26:29 ossec-logcollector: INFO: Monitoring full output of command(360): netstat -tulpn | sed 's/\([[:alnum:]]\+\)\ \+[[:digit:]]\+\ \+[[:digit:]]\+\ \+\(.*\):\([[:digit:]]*\)\ \+\([0-9\.\:\*]\+\).\+\ \([[:digit:]]*\/[[:alnum:]\-]*\).*/\1 \2 == \3 == \4 \5/' | sort -k 4 -g | sed 's/ == \(.*\) ==/:\1/' | sed 1,2d
2019/09/19 20:26:29 ossec-logcollector: INFO: Monitoring full output of command(360): last -n 20
2019/09/19 20:26:29 ossec-logcollector: INFO: Monitoring output of command(604800): /usr/bin/yum --disableplugin=versionlock --debuglevel 2 --security check-update 2&gt;/dev/null | grep -P '(?&lt;! 0 packages) available$'
2019/09/19 20:26:29 ossec-logcollector: ERROR: Couldn't stat file '/var/www/logs/access_log'
2019/09/19 20:26:29 ossec-logcollector: ERROR: Couldn't stat file '/var/www/logs/error_log'
2019/09/19 20:26:29 ossec-logcollector: ERROR: Couldn't stat file '/var/log/httpd/access_log.%Y%m%d-000000'
2019/09/19 20:26:29 ossec-logcollector: ERROR: Couldn't stat file '/var/log/httpd/error_log.%Y%m%d-000000'
2019/09/19 20:26:29 ossec-logcollector: ERROR: Couldn't stat file '/var/log/httpd/ssl_access_log.%Y%m%d-000000'
2019/09/19 20:26:29 ossec-logcollector: ERROR: Couldn't stat file '/var/log/httpd/ssl_request_log.%Y%m%d-000000'
2019/09/19 20:26:29 ossec-logcollector: ERROR: Couldn't stat file '/var/log/httpd/ssl_error_log.%Y%m%d-000000'
2019/09/19 20:26:29 ossec-logcollector: INFO: Started (pid: 19727).
2019/09/19 20:26:34 ossec-logcollector: INFO: Agent is now online. Process unlocked, continuing...
2019/09/19 20:27:38 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/log/messages'.
2019/09/19 20:27:38 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/log/authlog'.
2019/09/19 20:27:38 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/log/audit/audit.log'.
2019/09/19 20:27:38 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/log/secure'.
2019/09/19 20:27:38 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/log/xferlog'.
2019/09/19 20:27:38 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/log/maillog'.
2019/09/19 20:27:38 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/www/logs/access_log'.
2019/09/19 20:27:38 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/www/logs/error_log'.
2019/09/19 20:27:38 ossec-logcollector: INFO: (1952): Monitoring variable log file: '/var/log/httpd/access_log.20190919-000000'.
2019/09/19 20:27:38 ossec-logcollector: INFO: (1952): Monitoring variable log file: '/var/log/httpd/error_log.20190919-000000'.
2019/09/19 20:27:38 ossec-logcollector: INFO: (1952): Monitoring variable log file: '/var/log/httpd/ssl_access_log.20190919-000000'.
2019/09/19 20:27:38 ossec-logcollector: INFO: (1952): Monitoring variable log file: '/var/log/httpd/ssl_request_log.20190919-000000'.
2019/09/19 20:27:38 ossec-logcollector: INFO: (1952): Monitoring variable log file: '/var/log/httpd/ssl_error_log.20190919-000000'.
2019/09/19 20:35:19 ossec-logcollector: ERROR: Couldn't stat file '/var/log/authlog'
2019/09/19 20:35:19 ossec-logcollector: ERROR: Couldn't stat file '/var/log/authlog'
2019/09/19 20:35:19 ossec-logcollector: ERROR: Couldn't stat file '/var/log/authlog'
2019/09/19 20:35:19 ossec-logcollector: ERROR: Couldn't stat file '/var/log/authlog'
2019/09/19 20:35:19 ossec-logcollector: ERROR: Couldn't stat file '/var/log/xferlog'
2019/09/19 20:35:19 ossec-logcollector: ERROR: Couldn't stat file '/var/log/authlog'
2019/09/19 20:35:19 ossec-logcollector: INFO: Monitoring output of command(360): df -P
2019/09/19 20:35:19 ossec-logcollector: INFO: Monitoring full output of command(360): netstat -tulpn | sed 's/\([[:alnum:]]\+\)\ \+[[:digit:]]\+\ \+[[:digit:]]\+\ \+\(.*\):\([[:digit:]]*\)\ \+\([0-9\.\:\*]\+\).\+\ \([[:digit:]]*\/[[:alnum:]\-]*\).*/\1 \2 == \3 == \4 \5/' | sort -k 4 -g | sed 's/ == \(.*\) ==/:\1/' | sed 1,2d
2019/09/19 20:35:19 ossec-logcollector: INFO: Monitoring full output of command(360): last -n 20
2019/09/19 20:35:19 ossec-logcollector: INFO: Monitoring output of command(604800): /usr/bin/yum --disableplugin=versionlock --debuglevel 2 --security check-update 2&gt;/dev/null | grep -P '(?&lt;! 0 packages) available$'
2019/09/19 20:35:19 ossec-logcollector: ERROR: Couldn't stat file '/var/www/logs/access_log'
2019/09/19 20:35:19 ossec-logcollector: ERROR: Couldn't stat file '/var/www/logs/error_log'
2019/09/19 20:35:19 ossec-logcollector: ERROR: Couldn't stat file '/var/log/httpd/access_log.%Y%m%d-000000'
2019/09/19 20:35:19 ossec-logcollector: ERROR: Couldn't stat file '/var/log/httpd/error_log.%Y%m%d-000000'
2019/09/19 20:35:19 ossec-logcollector: ERROR: Couldn't stat file '/var/log/httpd/ssl_access_log.%Y%m%d-000000'
2019/09/19 20:35:19 ossec-logcollector: ERROR: Couldn't stat file '/var/log/httpd/ssl_request_log.%Y%m%d-000000'
2019/09/19 20:35:19 ossec-logcollector: INFO: Started (pid: 23701).
2019/09/19 20:35:24 ossec-logcollector: INFO: Agent is now online. Process unlocked, continuing...
2019/09/19 20:36:28 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/log/messages'.
2019/09/19 20:36:28 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/log/authlog'.
2019/09/19 20:36:28 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/log/audit/audit.log'.
2019/09/19 20:36:28 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/log/secure'.
2019/09/19 20:36:28 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/log/xferlog'.
2019/09/19 20:36:28 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/log/maillog'.
2019/09/19 20:36:28 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/www/logs/access_log'.
2019/09/19 20:36:28 ossec-logcollector: INFO: (1904): File not available, ignoring it: '/var/www/logs/error_log'.
2019/09/19 20:36:28 ossec-logcollector: INFO: (1952): Monitoring variable log file: '/var/log/httpd/access_log.20190919-000000'.
2019/09/19 20:36:28 ossec-logcollector: INFO: (1952): Monitoring variable log file: '/var/log/httpd/error_log.20190919-000000'.
2019/09/19 20:36:28 ossec-logcollector: INFO: (1952): Monitoring variable log file: '/var/log/httpd/ssl_access_log.20190919-000000'.
2019/09/19 20:36:28 ossec-logcollector: INFO: (1952): Monitoring variable log file: '/var/log/httpd/ssl_request_log.20190919-000000'.
2019/09/19 20:36:28 ossec-logcollector: INFO: (1952): Monitoring variable log file: '/var/log/httpd/ssl_error_log.20190919-000000'.
```

**Comment by vikman90** [comment]
@ssigwart thank you for your logs. We have implemented the fix and published a pull request (#3995).

Let us do some testing and approve it. We will definitely release this patch early next week.

Thank you again.
Best regards.

**Comment by vikman90** [comment]
@ssigwart We've just released Wazuh 3.10.2, including the patch for this issue.

Hope this helps you.

Let me close this issue. Again, thank you very much for your help with this.

Best regards.

**Comment by ssigwart** [comment]
Thanks for fixing that so quickly, @vikman90.  It seems to be working well now.
